File: sdk/python/cap/middleware.py
"""Middleware support for CAP agents and workers."""

import logging
import time
from typing import Any, Awaitable, Callable

from cap.runtime import Context

# NextFn calls the next middleware or the terminal handler.
NextFn = Callable[[Context, Any], Awaitable[Any]]

# Middleware intercepts handler execution. Call ``next_fn(ctx, data)`` to
# invoke the next middleware or the terminal handler. Middleware is applied
# in FIFO order: the first registered middleware is the outermost.
Middleware = Callable[[Context, Any, NextFn], Awaitable[Any]]


def logging_middleware(logger: logging.Logger = None) -> "Middleware":
    """Return a middleware that logs job ID, topic, and duration.

    Args:
        logger: Logger to use. Defaults to ``logging.getLogger("cap.middleware")``.
    """
    if logger is None:
        logger = logging.getLogger("cap.middleware")

    async def middleware(ctx: Context, data: Any, next_fn: NextFn) -> Any:
        start = time.monotonic()
        try:
            result = await next_fn(ctx, data)
            elapsed = int((time.monotonic() - start) * 1000)
            logger.info(
                "middleware: job=%s topic=%s duration=%dms ok",
                ctx.job_id,
                ctx.job.topic,
                elapsed,
            )
            return result
        except Exception as exc:
            elapsed = int((time.monotonic() - start) * 1000)
            logger.info(
                "middleware: job=%s topic=%s duration=%dms error=%s",
                ctx.job_id,
                ctx.job.topic,
                elapsed,
                exc,
            )
            raise

    return middleware

File: sdk/go/runtime/middleware.go
package runtime

import (
	"log"
	"time"
)

// HandlerFunc is the type-erased handler signature used internally by the Agent.
type HandlerFunc func(Context, any) (any, error)

// Middleware wraps a HandlerFunc, returning a new HandlerFunc.
// Middleware is applied in FIFO order: the first registered middleware is the outermost.
type Middleware func(next HandlerFunc) HandlerFunc

// Use appends middleware to the Agent. Middleware executes in registration order
// before the handler. Each middleware can inspect/modify the context, measure
// timing, or short-circuit by returning without calling next.
func (a *Agent) Use(mw ...Middleware) {
	a.middlewares = append(a.middlewares, mw...)
}

// applyMiddleware wraps handler through the middleware chain in FIFO order.
func applyMiddleware(handler HandlerFunc, mws []Middleware) HandlerFunc {
	h := handler
	for i := len(mws) - 1; i >= 0; i-- {
		h = mws[i](h)
	}
	return h
}

// LoggingMiddleware returns a Middleware that logs the job ID, topic, and
// execution duration for every request.
func LoggingMiddleware(logger *log.Logger) Middleware {
	return func(next HandlerFunc) HandlerFunc {
		return func(ctx Context, data any) (any, error) {
			start := time.Now()
			out, err := next(ctx, data)
			elapsed := time.Since(start)
			if err != nil {
				logger.Printf("middleware: job=%s topic=%s duration=%s error=%v",
					ctx.Job.GetJobId(), ctx.Job.GetTopic(), elapsed, err)
			} else {
				logger.Printf("middleware: job=%s topic=%s duration=%s ok",
					ctx.Job.GetJobId(), ctx.Job.GetTopic(), elapsed)
			}
			return out, err
		}
	}
}
